### PR TITLE
Social: Refactor the TemplatePicker component

### PR DESCRIPTION
**Pull request by gmjuhasz**
This change refactors our current `TemplatePicker` component, so that the inner logic can be used separately, decoupled from the modal. This is needed so we can use the picker in the SIG modal for #31665 .

## Proposed changes:

* Created new component for the picker itself. It has a `onTemplateSelected` prop, which can be used by the parent component to commit the changes, or do something with them.
* Used this new component inside our old one, which was renamed to `TemplatePickerModal`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

1204930070946424-as-1204986292229845
## Does this pull request change what data or activity we track or use?

No.
## Testing instructions:

* Open a new post with the Advanced plan.
  * Check that the template picker works as before:
    * It should open up correctly, and select the first - default - template.
    * Changing a template and saving should be reflected on the preview.
    * Re-opening the picker should show the selected template.
    * Publishing the post with SIG should use the correct image.
 * On the admin page, the template selector and changing default templates should work too.

### CODE REVIEW

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss [hunk 1]
```diff
@@ -0,0 +1,8 @@
+.footer {
+	display: flex;
+	justify-content: flex-end;
+
+	> * + * {
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss:5]: I know this is copied over from the other CSS file, but I personally don't like magic selectors like this. Could we use
```suggestion
	> :not(:first-child) {
```

to make it a bit clearer?

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js [hunk 2]
```diff
@@ -1,80 +1,58 @@
...
-	}, [ onSelect, setIsOpen, selectedTemplate ] );
-
 	// Add memoized callback function for each template.
 	const TEMPLATES = useMemo(
 		() =>
 			TEMPLATES_DATA.map( template => ( {
 				...template,
-				onSelect: () => setSelectedTemplate( template.name ),
+				onSelect: () => {
+					setSelectedTemplate( template.name );
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: Do we need this internal state anymore? The modal also has it, and I guess the new SIG settings modal would have it too.

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js [hunk 3]
```diff
@@ -1,80 +1,58 @@
...
+					if ( onTemplateSelected ) {
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: I wonder if there is a more succinct way to do this, especially if we remove the internal state.

Perhaps we could do something like:

```js
onSelect: () => onSelectTemplate?.( template.name )
```

I don't know if that's better or not!

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js [hunk 4]
```diff
@@ -1,80 +1,58 @@
...
-							{ __( 'Save', 'jetpack' ) }
-						</Button>
-					</div>
-				</Modal>
-			) }
-		</ThemeProvider>
+		<div className={ styles.templates }>
+			{ TEMPLATES.map( template => (
+				<button
+					onClick={ template.onSelect }
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: We might even be able to define the callback here then, so we don't have to set `onSelect` on each of the templates first.

**Review by pablinos — COMMENTED** [review]
This is working well for me. I've added a few comments to the code, and I think there are some versions that need updating after the weekly release, but it's pretty close.

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss (hunk 1 above)
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss:5]: Yeah sounds good. When we have the time we could clear up all these across social

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js (hunk 2 above)
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: Yeah probably we can get rid of it. It makes the TemplatePicker completely stateless though, so it always needs a parent that handles the state changes and rerenders it

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
-import { useState, useMemo } from '@wordpress/element';
 import { sprintf, __ } from '@wordpress/i18n';
 import classnames from 'classnames';
-import React from 'react';
+import React, { useCallback } from 'react';
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: I think this should come from `@wordpress/element`. I've seen problems where we end up importing multiple versions of React when we don't use the wrapped version. 

Similarly, we probably shouldn't importing `React` here. Is that for the JSDoc?

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js [hunk 6]
```diff
@@ -16,31 +15,23 @@ import TEMPLATES_DATA from './templates.js';
...
-					setSelectedTemplate( template.name );
-					if ( onTemplateSelected ) {
-						onTemplateSelected( template.name );
-					}
-				},
-			} ) ),
-		[ setSelectedTemplate, onTemplateSelected ]
+	const onTemplateClicked = useCallback(
+		event => {
+			const templateName = event.target.id;
```
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js:22]: Are we using a const to make it explicit that we're looking for the template's name?

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss (hunk 1 above)
Comment (pablinos) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker-modal/styles.module.scss:5]: Oh do we use this a lot? I think maybe this is a question of personal choice. The con to this is that it's a fair bit longer. Do you have a preference?

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js (hunk 5 above)
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: Yes for the JSDoc type

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js (hunk 6 above)
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js:22]: I think it makes it more readable, than just `onTemplateSelected?.( event.target.id );`. I could change it to an inline version too if you prefer that

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js (hunk 5 above)
Comment (pablinos) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js]: Seems like this should be solved in the `@wordpress/element` package, but [I notice this is handled elsewhere using a `@typedef`](https://github.com/WordPress/gutenberg/blob/e7b7f01b272c955ca51efb2f602a7ea16406f084/packages/block-editor/src/components/block-context/index.js#L6) although that is still importing from 'react'. What do you think doing it this way?

File: projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js (hunk 6 above)
Comment (pablinos) [reply projects/js-packages/publicize-components/src/components/social-image-generator/template-picker/index.js:22]: I don't have strong feelings really. I was more curious. It's fine to leave it as is.

**Comment by pablinos** [comment]
Just a thought about the ReactNode definition and we should probably get the E2E tests passing (or at least some of them!), but otherwise I think this is good.

**Review by pablinos — APPROVED** [review]
I think this is good now. I like that the component is now much simpler.

**Pull request merged**